src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -1518,29 +1510,72 @@ uct_ib_mlx5_devx_reg_exported_key(uct_ib_md_t *ib_md, uct_ib_mem_t *ib_memh)
+
+err_umem_dereg:
+    mlx5dv_devx_umem_dereg(memh->umem);
+    memh->umem = NULL;
+err_out:
+    return status;
+#else
+   return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_reg_exported_key(uct_ib_md_t *ib_md, uct_ib_mem_t *ib_memh)
+{
+    uct_ib_mlx5_md_t *md    = ucs_derived_of(ib_md, uct_ib_mlx5_md_t);
+    uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    uct_ib_mlx5_mr_t *mr    = &memh->mrs[UCT_IB_MR_DEFAULT];
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(allow_other_vhca_access_in)]   = {0};
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(allow_other_vhca_access_out)] = {0};
+    ucs_status_t status;
+    void *access_key;
+
+    ucs_assertv(memh->super.exported_lkey == UCT_IB_INVALID_MKEY,
+                "memh %p exported_lkey 0x%" PRIx32, memh,
+                memh->super.exported_lkey);
+again:
+    if (md->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI) {
+        status = uct_ib_mlx5_devx_reg_ksm_data_contig(md, mr, 0, 0,
+                &memh->cross_mr, &memh->super.exported_lkey);
+    } else {
+        status = uct_ib_mlx5_devx_create_umem_mr(md, memh);
+    }
+
+    if (status != UCS_OK) {
+        goto err_out;
+    }
 
-    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, ein, opcode,
+    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, in, opcode,
                       UCT_IB_MLX5_CMD_OP_ALLOW_OTHER_VHCA_ACCESS);
-    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, ein,
+    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, in,
                       object_type_to_be_accessed, UCT_IB_MLX5_OBJ_TYPE_MKEY);
-    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, ein,
+    UCT_IB_MLX5DV_SET(allow_other_vhca_access_in, in,
                       object_id_to_be_accessed,
                       memh->super.exported_lkey >> 8);
-    access_key = UCT_IB_MLX5DV_ADDR_OF(allow_other_vhca_access_in, ein,
+    access_key = UCT_IB_MLX5DV_ADDR_OF(allow_other_vhca_access_in, in,
                                        access_key);
     ucs_strncpy_zero(access_key, uct_ib_mkey_token,
                      UCT_IB_MLX5DV_FLD_SZ_BYTES(alias_context, access_key));
 
-    status = uct_ib_mlx5_devx_general_cmd(md->super.dev.ibv_context, ein,
-                                          sizeof(ein), eout, sizeof(eout),
-                                          "ALLOW_OTHER_VHCA_ACCESS", 0);
+    status = uct_ib_mlx5_devx_general_cmd(md->super.dev.ibv_context, in,
+                                          sizeof(in), out, sizeof(out),
+                                          "ALLOW_OTHER_VHCA_ACCESS", 1);
     if (status != UCS_OK) {
+        if (md->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI) {
+            ucs_debug("%s: indirect cross gvmi not supported, fallback to DEVX UMEM",
+                      uct_ib_device_name(&md->super.dev));
+            mlx5dv_devx_obj_destroy(memh->cross_mr);
+            memh->cross_mr = NULL;
+            md->flags &= ~UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+            goto again;
```

Comment:
Generally, `goto` going back is not so good practice, as it creates error-prone and hardly maintainable code.
1. Is there any chance of getting stuck in a loop here?
2. Can refactor to eliminate this `goto`?